src/iceberg/avro/avro_schema_util.cc

```diff
@@ -773,4 +775,301 @@ Result<SchemaProjection> Project(const Schema& expected_schema,
+                                                  const MappedField& field) {
+  if (original_node->leaves() != 2) {
+    return InvalidSchema("Map type must have exactly two leaves");
+  }
+
+  auto new_map_node = std::make_shared<::avro::NodeMap>();
+
+  // For map types, we need to extract key and value field mappings from the nested
+  // mapping
+  if (!field.nested_mapping) {
+    return InvalidSchema("Map type requires nested mapping for key and value fields");
+  }
+
+  // For map types, use the first two fields from nested mapping as key and value
+  if (!field.nested_mapping || field.nested_mapping->fields().size() < 2) {
+    return InvalidSchema("Map type requires nested mapping with key and value fields");
+  }
+
+  const auto& key_mapped_field = field.nested_mapping->fields()[0];
+  const auto& value_mapped_field = field.nested_mapping->fields()[1];
+
+  if (!key_mapped_field.field_id || !value_mapped_field.field_id) {
+    return InvalidSchema("Map key and value fields must have field IDs");
+  }
+
+  // Add key and value nodes
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto new_key_node,
+      MakeAvroNodeWithFieldIds(original_node->leafAt(0), key_mapped_field));
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto new_value_node,
+      MakeAvroNodeWithFieldIds(original_node->leafAt(1), value_mapped_field));
+  new_map_node->addLeaf(new_key_node);
+  new_map_node->addLeaf(new_value_node);
+
+  // Create key and value attributes
+  ::avro::CustomAttributes key_attributes;
+  ::avro::CustomAttributes value_attributes;
+
+  // Add required field IDs
+  key_attributes.addAttribute(std::string(kKeyIdProp),
+                              std::to_string(*key_mapped_field.field_id),
+                              /*addQuote=*/false);
+  value_attributes.addAttribute(std::string(kValueIdProp),
+                                std::to_string(*value_mapped_field.field_id),
+                                /*addQuote=*/false);
+
+  // Merge custom attributes from original node if they exist
+  if (original_node->customAttributes() > 0) {
+    // Merge attributes for key (index 0)
+    const auto& original_key_attrs = original_node->customAttributesAt(0);
+    const auto& existing_key_attrs = original_key_attrs.attributes();
+    for (const auto& attr_pair : existing_key_attrs) {
+      // Skip if it's the key ID property we're already setting
+      if (attr_pair.first != kKeyIdProp) {
+        key_attributes.addAttribute(attr_pair.first, attr_pair.second,
+                                    /*addQuote=*/false);
+      }
+    }
+
+    // Merge attributes for value (index 1)
+    if (original_node->customAttributes() > 1) {
+      const auto& original_value_attrs = original_node->customAttributesAt(1);
+      const auto& existing_value_attrs = original_value_attrs.attributes();
+      for (const auto& attr_pair : existing_value_attrs) {
+        // Skip if it's the value ID property we're already setting
+        if (attr_pair.first != kValueIdProp) {
```

Comment:
ditto.

BTW, do you think it worths adding a function to copy all custom attributes to a new one to avoid these repeated lines?